File: node/ts/net/chat/UnauthUsernamesService.ts
//
// Copyright 2025 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

import { Buffer } from 'node:buffer';
import * as uuid from 'uuid';

import * as Native from '../../Native.js';
import { Aci } from '../../Address.js';
import { Uuid } from '../../uuid.js';
import { RequestOptions, UnauthenticatedChatConnection } from '../Chat.js';

// For documentation
import type * as usernames from '../../usernames.js';
import type {
  InvalidEntropyDataLength,
  InvalidUsernameLinkEncryptedData,
} from '../../Errors.js';

declare module '../Chat' {
  // eslint-disable-next-line @typescript-eslint/no-empty-object-type
  interface UnauthenticatedChatConnection extends UnauthUsernamesService {}
}

export interface UnauthUsernamesService {
  /**
   * Looks up a username hash on the service, like that computed by {@link usernames.hash}.
   *
   * Returns the corresponding account's ACI, or `null` if the username doesn't correspond to an
   * account.
   *
   * Throws / completes with failure only if the request can't be completed, potentially including
   * if the hash is structurally invalid.
   */
  lookUpUsernameHash: (
    request: {
      hash: Uint8Array;
    },
    options?: RequestOptions
  ) => Promise<Aci | null>;

  /**
   * Looks up a username link on the service by UUID.
   *
   * Returns a decrypted, validated username, or `null` if the UUID does not correspond to a
   * username link (perhaps the user rotated their link).
   *
   * Throws / completes with failure if the request can't be completed. Specifically throws
   * {@link InvalidEntropyDataLength} if the entropy is invalid, and
   * {@link InvalidUsernameLinkEncryptedData} if the data fetched from the service could not be
   * decrypted or did not contain a valid username. `uuid` should be validated ahead of time.
   */
  lookUpUsernameLink: (
    request: {
      uuid: Uuid;
      entropy: Uint8Array;
    },
    options?: RequestOptions
  ) => Promise<{ username: string; hash: Uint8Array } | null>;
}

UnauthenticatedChatConnection.prototype.lookUpUsernameHash = async function (
  {
    hash,
  }: {
    hash: Uint8Array;
  },
  options?: RequestOptions
): Promise<Aci | null> {
  const response = await this._asyncContext.makeCancellable(
    options?.abortSignal,
    Native.UnauthenticatedChatConnection_look_up_username_hash(
      this._asyncContext,
      this._chatService,
      Buffer.from(hash)
    )
  );
  return response ? Aci.fromUuidBytes(response) : null;
};

UnauthenticatedChatConnection.prototype.lookUpUsernameLink = async function (
  {
    uuid: linkUuid,
    entropy,
  }: {
    uuid: Uuid;
    entropy: Uint8Array;
  },
  options?: RequestOptions
): Promise<{ username: string; hash: Uint8Array } | null> {
  const response = await this._asyncContext.makeCancellable(
    options?.abortSignal,
    Native.UnauthenticatedChatConnection_look_up_username_link(
      this._asyncContext,
      this._chatService,
      uuid.parse(linkUuid),
      entropy
    )
  );
  return response ? { username: response[0], hash: response[1] } : null;
};

File: swift/Sources/LibSignalClient/chat/UnauthUsernamesService.swift
//
// Copyright 2025 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

import Foundation
import SignalFfi

public protocol UnauthUsernamesService: Sendable {
    /// Looks up a username hash on the service, like that computed by ``Username``.
    ///
    /// Returns the corresponding account's ACI, or `nil` if the username doesn't correspond to an
    /// account.
    ///
    /// Throws only if the request can't be completed, potentially including if the hash is
    /// structurally invalid.
    func lookUpUsernameHash(_ hash: Data) async throws -> Aci?

    /// Looks up a username link on the service by UUID.
    ///
    /// Returns a decrypted, validated username, or `nil` if the UUID does not correspond to a
    /// username link (perhaps the user rotated their link).
    ///
    /// Throws if the request can't be completed. Specifically throws
    /// ``SignalError/usernameLinkInvalidEntropyDataLength(_:)`` if the entropy is invalid, and
    /// ``SignalError/usernameLinkInvalid(_:)`` if the data fetched from the service could not be
    /// decrypted or did not contain a valid username.
    func lookUpUsernameLink(_ uuid: UUID, entropy: Data) async throws -> Username?
}

extension UnauthenticatedChatConnection: UnauthUsernamesService {
    public func lookUpUsernameHash(_ hash: Data) async throws -> Aci? {
        let rawResponse: SignalOptionalUuid = try await self.tokioAsyncContext
            .invokeAsyncFunction { promise, tokioAsyncContext in
                withNativeHandle { chatService in
                    hash.withUnsafeBorrowedBuffer { hash in
                        signal_unauthenticated_chat_connection_look_up_username_hash(
                            promise,
                            tokioAsyncContext.const(),
                            chatService.const(),
                            hash
                        )
                    }
                }
            }
        let uuid = try! invokeFnReturningOptionalUuid { out in
            out?.pointee = rawResponse
            return nil
        }
        return uuid.map { Aci(fromUUID: $0) }
    }

    public func lookUpUsernameLink(_ uuid: UUID, entropy: Data) async throws -> Username? {
        var rawResponse: SignalOptionalPairOfc_charu832 = try await self.tokioAsyncContext
            .invokeAsyncFunction { promise, tokioAsyncContext in
                withNativeHandle { chatService in
                    try! withAllBorrowed(uuid, entropy) { uuid, entropy in
                        signal_unauthenticated_chat_connection_look_up_username_link(
                            promise,
                            tokioAsyncContext.const(),
                            chatService.const(),
                            uuid,
                            entropy
                        )
                    }
                }
            }
        guard rawResponse.present else {
            return nil
        }
        defer { signal_free_string(rawResponse.first) }
        let name = String(cString: rawResponse.first!)
        let hash = withUnsafeBytes(of: &rawResponse.second) { Data($0) }
        return Username(name, uncheckedHash: hash)
    }
}

extension UnauthServiceSelector where Self == UnauthServiceSelectorHelper<any UnauthUsernamesService> {
    public static var usernames: Self { .init() }
}
